### PR TITLE
CDD-1763: Added batch file upload upscan callback endpoint.

### PR DESCRIPTION
**Pull request by googley42**

### CODE REVIEW

File: app/uk/gov/hmrc/customs/declaration/services/BatchFileUploadUpscanNotificationBusinessService.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        cb.name,
+        cb.mimeType,
+        cb.checksum,
+        bf.location,
+        bf.sequenceNumber
+      )
+    )
+  }
+
+}
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/BatchFileUploadUpscanNotificationBusinessService.scala]: General comment on above - my personal preference is to not use this many abbreviations. bf, ftf, md, cb... Makes reading the code more difficult than it needs to be. But, if you prefer this...

File: app/uk/gov/hmrc/customs/declaration/services/InternalErrorXmlNotification.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import scala.xml.NodeSeq
+
+@Singleton
+class InternalErrorXmlNotification extends CallbackToXmlNotification[FileReference] {
+
+  override def toXml(fileReference: FileReference): NodeSeq =
+    <errorResponse>
+      <code>INTERNAL_SERVER_ERROR</code>
+      <message>File upload for file reference {fileReference.toString} failed. A system error has prevented your document from being  accepted. Please follow the guidance on www.gov.uk and submit your documents by an alternative method.</message>
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/customs/declaration/services/InternalErrorXmlNotification.scala]: unnecessary space between `being` and `accepted`

File: test/component/BatchFileUploadUpscanNotificationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      Given("the Upscan Initiate service has been sent a valid request with this Declaration API as the callback URL")
+
+      When("Upscan Notify service notifies this Declaration API using previously provided callback URL")
+      val validRequest = FakeRequest("POST", endpoint).withJsonBody(readyJson(FileReferenceTwo))
+      val result = route(app = app, validRequest).value
+
+      Then("Declaration Service returns 500")
+      status(result) shouldBe 500
+
+      And("the response body is empty")
```
Comment (paulblakehmrc) [inline test/component/BatchFileUploadUpscanNotificationSpec.scala]: `empty` should be replaced with correct response

File: test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+      actual shouldBe (())
+      verify(mockRepo).update(
+        ameq[UUID](FileReferenceOne.value).asInstanceOf[FileReference],
+        ameq(callbackFields))(any[HasConversationId]
+      )
+      verify(mockConnector).send(ameq(fileTransmissionRequest))
+    }
+
+    "return failed future when no metatdata record found for file reference" in new SetUp {
```
Comment (paulblakehmrc) [inline test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala]: `metatdata` - typo

File: test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+      error.getMessage shouldBe s"database error - can't find record with file reference ${FileReferenceOne.value.toString}"
+      verify(mockRepo).update(
+        ameq[UUID](FileReferenceOne.value).asInstanceOf[FileReference],
+        ameq(callbackFields))(any[HasConversationId]
+      )
+      verifyZeroInteractions(mockConnector)
+    }
+
+    "return failed future when file reference not found in returned metatdata" in new SetUp {
```
Comment (paulblakehmrc) [inline test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala]: `metatdata` - typo

File: app/uk/gov/hmrc/customs/declaration/connectors/FileTransmissionConnector.scala [hunk 6]
```diff
@@ -37,7 +37,7 @@ class FileTransmissionConnector @Inject()(http: HttpClient,
     extraHeaders = Seq(ACCEPT -> JSON, CONTENT_TYPE -> JSON, USER_AGENT -> "customs-declarations")
   )
 
-  def send[A](request: FileTransmission, apiVersion: ApiVersion): Future[Unit] = {
+  def send[A](request: FileTransmission): Future[Unit] = {
```
[the remark this answers is not in the document]
Comment (googley42) [reply app/uk/gov/hmrc/customs/declaration/connectors/FileTransmissionConnector.scala]: done

File: app/uk/gov/hmrc/customs/declaration/controllers/BatchFileUploadUpscanNotificationController.scala [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  private def internalServerErrorResult(e: Throwable)(implicit request: Request[AnyContent]) = {
+    cdsLogger.error(s"Error processing file transmission callback request. Body: ${request.body.asText} headers: ${request.headers}", e)
+    ErrorInternalServerError.JsonResult
+  }
+
+  private def asyncNotifyInternalServerError(callbackBody: UploadedCallbackBody, subscriptionFieldsId: SubscriptionFieldsId)(implicit request: Request[AnyContent]) = {
+    //TODO: Use persistent retry when this is available
+    Future {
+      (notificationService.sendMessage[FileReference](
+        callbackBody.reference,
```
[the remark this answers is not in the document]
Comment (googley42) [reply app/uk/gov/hmrc/customs/declaration/controllers/BatchFileUploadUpscanNotificationController.scala]: done

File: app/uk/gov/hmrc/customs/declaration/controllers/BatchFileUploadUpscanNotificationController.scala [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                cdsLogger.debug(s"Valid JSON request received. Body: $js headers: ${request.headers}")
+                businessService.persistAndCallFileTransmission(ready).map{_ =>
+                    Results.NoContent
+                }.recover{
+                  case e: Throwable =>
+                    asyncNotifyInternalServerError(ready, clientSubscriptionId)
+                    internalServerErrorResult(e)
+                }
+              case failed: UploadedFailedCallbackBody =>
+                notificationService.sendMessage[UploadedFailedCallbackBody](
```
[the remark this answers is not in the document]
Comment (googley42) [reply app/uk/gov/hmrc/customs/declaration/controllers/BatchFileUploadUpscanNotificationController.scala]: done

File: app/uk/gov/hmrc/customs/declaration/services/BatchFileUploadUpscanNotificationBusinessService.scala (hunk 1 above)
Comment (googley42) [reply app/uk/gov/hmrc/customs/declaration/services/BatchFileUploadUpscanNotificationBusinessService.scala]: done

File: app/uk/gov/hmrc/customs/declaration/services/InternalErrorXmlNotification.scala (hunk 2 above)
Comment (googley42) [reply app/uk/gov/hmrc/customs/declaration/services/InternalErrorXmlNotification.scala]: done

File: test/component/BatchFileUploadUpscanNotificationSpec.scala (hunk 3 above)
Comment (googley42) [reply test/component/BatchFileUploadUpscanNotificationSpec.scala]: done

File: test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala (hunk 4 above)
Comment (googley42) [reply test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala]: done

File: test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala (hunk 5 above)
Comment (googley42) [reply test/unit/services/BatchFileUploadUpscanNotificationBusinessServiceSpec.scala]: done

**Pull request merged**